# Web Mining and Applied NLP (44-620)

## Requests, JSON, and NLP

### Student Name: Topaz Montague 

### Linke to Project Repository: [P4 GitHub Repository](https://github.com/tfmontague/json-sentiment)

Perform the tasks described in the Markdown cells below.  When you have completed the assignment make sure your code cells have all been run (and have output beneath them) and ensure you have committed and pushed ALL of your changes to your assignment repository.

Make sure you have [installed spaCy and its pipeline](https://spacy.io/usage#quickstart) and [spaCyTextBlob](https://spacy.io/universe/project/spacy-textblob)

Every question that requires you to write code will have a code cell underneath it; you may either write your entire solution in that cell or write it in a python file (`.py`), then import and run the appropriate code to answer the question.

This assignment requires that you write additional files (either JSON or pickle files); make sure to submit those files in your repository as well.

1. The following code accesses the [lyrics.ovh](https://lyricsovh.docs.apiary.io/#reference/0/lyrics-of-a-song/search) public api, searches for the lyrics of a song, and stores it in a dictionary object.  Write the resulting json to a file (either a JSON file or a pickle file; you choose). You will read in the contents of this file for future questions so we do not need to frequently access the API.

**Note**: The current instructions are no longer working. We will use: https://gist.github.com/neloe/e38e4d3283418e096aac80fa48bb66bd . Instead of artist/song, we will use author/poem to fetch our content.

#### Question 1 Response

In [13]:
import requests
import json

AUTHOR = 'Edgar Allan Poe'
POEM = 'A Dream Within A Dream'

URL = f'https://poetrydb.org/author,title/{AUTHOR};{POEM}'
result = json.loads(requests.get(URL).text)

# write the result to a JSON file
with open('poem.json', 'w') as file:
    json.dump(result, file)

# Print the poem to get results on the screen
poem = '\n'.join(result[0]['lines'])
print(poem)


Take this kiss upon the brow!
And, in parting from you now,
Thus much let me avow--
You are not wrong, who deem
That my days have been a dream:
Yet if hope has flown away
In a night, or in a day,
In a vision or in none,
Is it therefore the less _gone_?
_All_ that we see or seem
Is but a dream within a dream.

I stand amid the roar
Of a surf-tormented shore,
And I hold within my hand
Grains of the golden sand--
How few! yet how they creep
Through my fingers to the deep
While I weep--while I weep!
O God! can I not grasp
Them with a tighter clasp?
O God! can I not save
_One_ from the pitiless wave?
Is _all_ that we see or seem
But a dream within a dream?


2. Read in the contents of your file.  Print the lyrics of the song (not the entire dictionary!) and use spaCyTextBlob to perform sentiment analysis on the lyrics.  Print the polarity score of the sentiment analysis.  Given that the range of the polarity score is `[-1.0,1.0]` which corresponds to how positive or negative the text in question is, do you think the lyrics have a more positive or negative connotaion?  Answer this question in a comment in your code cell.

**Note**: The current instructions are no longer working. We will use: https://gist.github.com/neloe/e38e4d3283418e096aac80fa48bb66bd . Instead of artist/song, we will use author/poem to fetch our content.

#### Question 2 Response

In [14]:
import json
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

# Initialize spaCy
nlp = spacy.load('en_core_web_sm')
nlp.add_pipe('spacytextblob')

# Read in the contents of the JSON file
with open('poem.json', 'r') as file:
    result = json.load(file)

# Extract the poem's text (lyrics)
poem_lines = result[0]['lines']
poem_text = '\n'.join(poem_lines)

# Print the lyrics of the song (poem)
print("Poem Lyrics:\n", poem_text)

# Perform sentiment analysis on the lyrics
doc = nlp(poem_text)
polarity_score = doc._.polarity

# Print the polarity score of the sentiment analysis
print("\nPolarity Score:", polarity_score)

# Given that the polarity score of 0.055555555555555546 falls slightly above the midpoint of 0. This indicates that the text (poem lyrics) has a slightly positive connotation. The score is very close to neutral but leans towards the positive side, suggesting that the overall sentiment of the poem is marginally more positive than negative.

Poem Lyrics:
 Take this kiss upon the brow!
And, in parting from you now,
Thus much let me avow--
You are not wrong, who deem
That my days have been a dream:
Yet if hope has flown away
In a night, or in a day,
In a vision or in none,
Is it therefore the less _gone_?
_All_ that we see or seem
Is but a dream within a dream.

I stand amid the roar
Of a surf-tormented shore,
And I hold within my hand
Grains of the golden sand--
How few! yet how they creep
Through my fingers to the deep
While I weep--while I weep!
O God! can I not grasp
Them with a tighter clasp?
O God! can I not save
_One_ from the pitiless wave?
Is _all_ that we see or seem
But a dream within a dream?

Polarity Score: 0.055555555555555546


3. Write a function that takes an artist, song, and filename, accesses the lyrics.ovh api to get the song lyrics, and writes the results to the specified filename.  Test this function by getting the lyrics to any four songs of your choice and storing them in different files.

**Note**: The current instructions are no longer working. We will use: https://gist.github.com/neloe/e38e4d3283418e096aac80fa48bb66bd . Instead of artist/song, we will use author/poem to fetch our content.

#### Question 3 Response

In [15]:
import requests
import json

def fetch_and_save_poem(author, title, filename):
    """
    Fetches a poem given an author and title from the PoetryDB API and saves the result to a specified filename.

    Parameters:
    - author: The name of the author.
    - title: The title of the poem.
    - filename: The filename where the poem will be saved.
    """
    url = f'https://poetrydb.org/author,title/{author};{title}'
    response = requests.get(url)
    if response.status_code == 200:
        data = response.json()
        with open(filename, 'w') as file:
            json.dump(data, file)
        print(f"Saved '{title}' by {author} to {filename}")
    else:
        print(f"Failed to fetch '{title}' by {author}. Status code: {response.status_code}")

# Testing the function with the specified poems
fetch_and_save_poem("Robert Frost", "The Road Not Taken", "the_road_not_taken.json")
fetch_and_save_poem("Robert Frost", "Stopping by Woods on a Snowy Evening", "stopping_by_woods_on_a_snowy_evening.json")
fetch_and_save_poem("William Shakespeare", "Sonnet 18", "sonnet_18.json")
fetch_and_save_poem("Emily Dickinson", "Because I could not stop for Death", "because_i_could_not_stop_for_death.json")


Saved 'The Road Not Taken' by Robert Frost to the_road_not_taken.json
Saved 'Stopping by Woods on a Snowy Evening' by Robert Frost to stopping_by_woods_on_a_snowy_evening.json
Saved 'Sonnet 18' by William Shakespeare to sonnet_18.json
Saved 'Because I could not stop for Death' by Emily Dickinson to because_i_could_not_stop_for_death.json


4. a. Write a function that takes the name of a file that contains song lyrics, loads the file, performs sentiment analysis, and returns the polarity score.  Use this function to print the polarity scores (with the name of the song) of the three files you created in question 3.  b. Does the reported polarity match your understanding of the song's lyrics? Why or why not do you think that might be?  Answer the questions in either a comment in the code cell or a markdown cell under the code cell.

**Note**: The current instructions are no longer working. We will use: https://gist.github.com/neloe/e38e4d3283418e096aac80fa48bb66bd . Instead of artist/song, we will use author/poem to fetch our content.

#### Question 4a Response

In [16]:

# Initialize spaCy and add the SpacyTextBlob pipeline
nlp = spacy.load("en_core_web_sm")
nlp.add_pipe('spacytextblob')

# Function to analyze sentiment and get polarity score
def analyze_sentiment_and_get_polarity(filename):
    with open(filename, 'r') as file:
        data = json.load(file)
        
        # Handling both possible JSON structures
        if isinstance(data, list):
            poem_lines = data[0]['lines']
        else:
            poem_lines = data['lines'] if 'lines' in data else []
            
        poem_text = '\n'.join(poem_lines)
        doc = nlp(poem_text)
        return doc._.polarity

# Initialize spaCy and add the SpacyTextBlob pipeline
nlp = spacy.load("en_core_web_sm")
nlp.add_pipe('spacytextblob')

# Fetch and save the poems
poems = [
    ("Robert Frost", "The Road Not Taken", "the_road_not_taken.json"),
    ("Robert Frost", "Stopping by Woods on a Snowy Evening", "stopping_by_woods_on_a_snowy_evening.json"),
    ("William Shakespeare", "Sonnet 18", "sonnet_18.json"),
    ("Emily Dickinson", "Because I could not stop for Death", "because_i_could_not_stop_for_death.json")
]

for author, title, filename in poems:
    fetch_and_save_poem(author, title, filename)

# Analyze sentiment and print polarity scores
for author, title, filename in poems:
    polarity_score = analyze_sentiment_and_get_polarity(filename)
    print(f"Polarity Score for '{title}': {polarity_score}")


Saved 'The Road Not Taken' by Robert Frost to the_road_not_taken.json
Saved 'Stopping by Woods on a Snowy Evening' by Robert Frost to stopping_by_woods_on_a_snowy_evening.json
Saved 'Sonnet 18' by William Shakespeare to sonnet_18.json
Saved 'Because I could not stop for Death' by Emily Dickinson to because_i_could_not_stop_for_death.json
Polarity Score for 'The Road Not Taken': 0.0
Polarity Score for 'Stopping by Woods on a Snowy Evening': 0.0
Polarity Score for 'Sonnet 18': 0.33181818181818185
Polarity Score for 'Because I could not stop for Death': 0.0


#### Question 4b Response

| Poem Title                           | Polarity Score        | Does Reported Polarity Match My Understanding? | Why or Why Not                                                                 |
|--------------------------------------|-----------------------|------------------------------------------------|--------------------------------------------------------------------------------|
| The Road Not Taken                   | 0.0                   | No                                             | Sentiment may not have picked up on underlying negative context                |
| Stopping by Woods on a Snowy Evening | 0.0                   | Yes                                            | There was an equal amount of positive and negative context                    |
| Sonnet 18                            | 0.33181818181818185   | Yes                                            | There was more positive than negative context                                  |
| Because I could not stop for Death   | 0.0                   | No                                             | Sentiment may not have picked up on underlying negative context                |
